lib/modules/manager/gradle/extract/consistent-versions-plugin.ts

```diff
@@ -0,0 +1,187 @@
+export const VERSIONS_PROPS = 'versions.props';
+export const VERSIONS_LOCK = 'versions.lock';
+
+/**
+ * Determines if Palantir Gradle consistent-versions is in use, https://github.com/palantir/gradle-consistent-versions.
+ * The plugin name must be in build file and both `versions.props` and `versions.lock` must exist.
+ *
+ * @param availableFiles list of gradle build files found in project
+ * @param fileContents map with file contents of all files
+ */
+export function usesGcv(
+  availableFiles: string[],
+  fileContents: Record<string, string | null>
+): boolean {
+  if (
+    fileContents['build.gradle']?.includes(
+      'com.palantir.consistent-versions'
+    ) ||
+    fileContents['build.kts']?.includes('com.palantir.consistent-versions')
+  ) {
+    if (
+      availableFiles.includes(VERSIONS_PROPS) &&
+      availableFiles.includes(VERSIONS_LOCK)
+    ) {
+      logger.debug('This repo uses gradle-consistent-versions');
+      return true;
+    }
+  }
+  return false;
+}
+
+/**
+ * Confirms whether the provided file name is one of the two GCV files
+ */
+export function isGcvPropsFile(fileName: string): boolean {
+  return fileName === VERSIONS_PROPS || fileName === VERSIONS_LOCK;
+}
+
+/**
+ * Parses Gradle-Consistent-Versions files to figure out what dependencies, versions
+ * and groups they contain. The parsing goes like this:
+ * - Parse `versions.props` into deps (or groups) and versions, remembering file offsets
+ * - Parse `versions.lock` into deps and lock-versions
+ * - For each exact dep in props file, lookup the lock-version from lock file
+ * - For each group/regex dep in props file, lookup the set of exact deps and versions in lock file
+ *
+ * @param propsFileContent text content of `versions.props
+ * @param lockFileContent text content of `versions.lock`
+ */
+export function parseGcv(
+  propsFileContent: string,
+  lockFileContent: string
+): PackageDependency<GradleManagerData>[] {
+  const lockFileMap = parseLockFile(lockFileContent);
+  const [propsFileExactMap, propsFileRegexMap] =
+    parsePropsFile(propsFileContent);
+
+  const extractedDeps: PackageDependency<GradleManagerData>[] = [];
+
+  // For each exact dep in props file
+  for (const [propDep, versionAndPosition] of propsFileExactMap) {
+    if (lockFileMap.has(propDep)) {
+      const newDep: Record<string, any> = {
+        managerData: {
+          packageFile: VERSIONS_PROPS,
+          fileReplacePosition: versionAndPosition.filePos,
+        },
+        packageName: propDep,
+        currentValue: versionAndPosition.version,
+        currentVersion: versionAndPosition.version,
+        lockedVersion: lockFileMap.get(propDep),
+      } as PackageDependency<GradleManagerData>;
+      extractedDeps.push(newDep);
+    }
+  }
+
+  // For each regular expression (group) dep in props file
+  for (const [propDepRegEx, propVerAndPos] of propsFileRegexMap) {
+    for (const [exactDep, exactVer] of lockFileMap) {
+      if (propDepRegEx.test(exactDep)) {
```

Comment:
In addition to testing that we get a regex match, we should also make sure that the match does not exist as an excact (override) line in props file. Example props snippet:

```
org.example:foo = 2.0
org.example:* = 1.9
```

With the example above w want all `org.example` deps to use v1.9 except `foo` which should be on 2.0. However, as is now, both lines in props will match the same line in lock. The exact one shold always win.

Reply:
This is now implemented, and a test added for it.